### ISSUE TITLE
Validation issue with missing nationalPrefixOptionalWhenFormatting (AU)

### ISSUE THREAD

**Issue by onlineautoparts** [googlei18n/libphonenumber#1696]
Hello, 

Australia's metadata doesn't include "nationalPrefixOptionalWhenFormatting" which prevents the third party JavaScript port by @halt-hammerzeit from recognizing "413071810" as a valid phone number. It will only recognize the number when it includes a "0" i.e. "0413071810".
The Javascript port I'm using is https://github.com/halt-hammerzeit/libphonenumber-js. When testing "413071810" in your javascript demo it does say it is a valid number for the region of AU.

After creating an issue at @halt-hammerzeit repository (https://github.com/halt-hammerzeit/libphonenumber-js/issues/67). He suggests I bring this issue to you:

>@halt-hammerzeit I had a look at your case and Google's demo indeed says that this is a valid number.
I personally think Google's algorythm is wrong and if 0 is really not required then you may create an issue in their repo to add nationalPrefixOptionalWhenFormatting="true" either to the format itself to to the whole AU country.
https://github.com/googlei18n/libphonenumber/blob/3894ee185c277ee5d0c9e81993030aaaff4a8d50/resources/PhoneNumberMetadata.xml#L1725-L1732

**Comment by halt-hammerzeit** [comment]
For example, in Russia `nationalPrefixOptionalWhenFormatting` is `true` and we can both write it as `800 555 35 35` and `8 800 555 35 35`.
Either there is a bug in the metadata and `nationalPrefixOptionalWhenFormatting=true` must be added for Australia or the `nationalPrefixOptionalWhenFormatting` is completely meaningless.

**Comment by penmetsaa** [comment]
Hi,

Thanks for reporting! 

Libphonenumber is working as intended here. Both our Java and JS demos also are returning 413071810 (AU) as a valid number.

Java Demo: https://libphonenumber.appspot.com/phonenumberparser?number=413071810&country=AU
JS Demo: https://rawgit.com/googlei18n/libphonenumber/master/javascript/i18n/phonenumbers/demo-compiled.html

This needs to be fixed at halt-hammerzeit/libphonenumber-js port. Please contact back to them referring this info. 

```nationalPrefixOptionalWhenFormatting``` is not used for parsing the numbers. It is used only when formatting the numbers. More details in [phonemetadata.proto](https://github.com/googlei18n/libphonenumber/blob/a342989093fbd56c75b3cba655b3107d4d9a823a/resources/phonemetadata.proto#L87)

**Comment by halt-hammerzeit** [comment]
> nationalPrefixOptionalWhenFormatting is not used for parsing the numbers. 

Oh yeah?
You didn't read the CPP source code for sure
https://github.com/googlei18n/libphonenumber/blob/eab5571b7086f603a207129024152ce2c6d1e050/cpp/src/phonenumbers/phonenumbermatcher.cc#L769

> It is used only when formatting the numbers.

Oh rly?
Then show me the line of code where it is "used when formatting the numbers".
If you don't then you didn't read neither the Java or the Javascript source codes too.

**Comment by halt-hammerzeit** [comment]
I decided to remove support for the `nationalPrefixOptionalWhenFormatting` flag while parsing because its behaviour inconsistent during parsing in this original library.

**Comment by penmetsaa** [comment]
@halt-hammerzeit ,

>> This is still a bug

The original issue is about why input number 413071810 (having no national prefix) is considered as invalid in JS. However this is not reproducible at our JS / Java demo websites. https://rawgit.com/googlei18n/libphonenumber/master/javascript/i18n/phonenumbers/demo-compiled.html

>> You didn't read the CPP source code for sure

Thanks for pointing to this code. national_prefix_optional_when_formatting is also used for matching  two numbers but not during parsing / validation as mentioned earlier.

>> I decided to remove support for the nationalPrefixOptionalWhenFormatting flag while parsing because its behaviour is inconsistent between different language implementations during parsing in this original library.

Can you please share the inputs / demo links where the results are different in two languages. It helps us to solve if any inconsistencies are present.

**Comment by halt-hammerzeit** [comment]
@penmetsaa Okay then, you're right

**Comment by lararennie** [comment]
FYI: this field is set to true if we encounter numbers in the wild commonly
written without the national prefix, when we know the "right" way is to
include them. For Australia, we have no evidence that this is the case.
(This does not affect numbers which are never written with one.)

If we set it to true where this is not the case, our functionality to find
phone numbers in text (that you pointed to) suffers precision losses, and
marginal to none recall improvement.

We don't use it for parsing or validation so a true port wouldn't either.

Would it help if we expanded the comment in the proto to explain what we
use it for?

18. apr. 2017 2.13 p.m. skrev "Nikolay" <notifications@github.com>:

@penmetsaa <https://github.com/penmetsaa> Okay then, you're right

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/googlei18n/libphonenumber/issues/1696#issuecomment-294812512>,
or mute the thread
